### PR TITLE
chore: update a part of libanki

### PR DESCRIPTION
**Pull request by BrayanDSO**
## Purpose / Description

I just looked into a method and started updating it like a robot. When the spell was gone, I stopped it.

## How Has This Been Tested?

I haven't

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt [hunk 1]
```diff
@@ -461,7 +461,21 @@ class Decks(private val col: Collection) {
      *************************************************************
      */
 
-    // TODO
+    /** The deck of did and all its children, as (name, id). */
+    @LibAnkiAlias("deck_and_child_name_ids")
+    fun deckAndChildNameIds(deckId: DeckId): List<Pair<String, Long>> {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt:466]: this matches upstream: `Iterable[tuple[str, DeckId]]:`

I'd prefer to use `DeckNameId` here. Implementer's choice

File: AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt [hunk 2]
```diff
@@ -515,6 +515,27 @@ class Decks(private val col: Collection) {
         return parents.toList()
     }
 
+    /** All existing parents of name */
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt:518]: `[name]`

File: AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt [hunk 3]
```diff
@@ -515,6 +515,27 @@ class Decks(private val col: Collection) {
...
+        val head = mutableListOf<String>()
+        val parents = mutableListOf<Deck>()
+
+        for (deckName in names) {
+            head.add(deckName)
+            val deck = byName(head.joinToString("::"))
+            if (deck != null) {
+                parents.append(deck)
+            }
+        }
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt:534]: Reviewer 2- compare to https://github.com/ankitects/anki/blob/4fc535ec8810c77838d49252ed83921ff99301a6/pylib/anki/decks.py#L503-L517 as the code differs

File: AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt [hunk 4]
```diff
@@ -515,6 +515,27 @@ class Decks(private val col: Collection) {
         return parents.toList()
     }
 
+    /** All existing parents of name */
+    @LibAnkiAlias("parents_by_name")
+    fun parentsByName(name: String): List<Deck> {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt:520]: `@NeedsTest` as the code differs slightly

**Review by david-allison — CHANGES_REQUESTED** [review]
Wanted to do this myself, thanks!

Could you add `@CheckResult` to relevant functions, especially ones returning `opChanges`

File: AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt (hunk 1 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt:466]: I read `DeckNameId` and get more confused, so I'll just rename Long to DeckId

**Review by david-allison — APPROVED** [review]
The commit comments were AWESOME, thank you for that

File: AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt (hunk 3 above)
Comment (lukstbit) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Decks.kt:536]: It's different but the logic is the same in my opinion.

**Review by lukstbit — APPROVED** [review]
Great!

**Pull request merged**